[MM-20061] Add System Admin filter to both LDAP and SAML settings in Admin Console

hahmadia opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

**Merge Third**

#### Summary
<!--
A description of what this pull request does.
-->
Adds the front end portion for system admin filter in LDAP and SAML

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
https://mattermost.atlassian.net/browse/MM-20061

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes 
1. https://github.com/mattermost/mattermost-server/pull/13534  
2. https://github.com/mattermost/enterprise/pull/557

sbishel commented on components/admin_console/admin_definition.jsx:
```diff
@@ -2454,6 +2454,21 @@ const AdminDefinition = {
+                        label: t('admin.ldap.adminFilterTitle'),
+                        label_default: 'Admin Filter:',
+                        help_text: t('admin.ldap.adminFilterFilterDesc'),
+                        help_text_default: '(Optional) Enter an AD/LDAP filter to use when searching for admin objects. Only the users selected by the query will be able to access Mattermost as Admins.\n \nNote: If this filter is removed/changed, active admins will not be demoted to a member and will retain their Admin role. Admins can be demoted in **System Console > User Management**.\n \n \nEx ...
```
"Note: If this filter is removed/changed, active admins will not be demoted to a member and will retain their Admin role." 
Based on prior review this isn't true. They are being promoted and demoted on next login. 
@thefactremains - Can you take a look at this text and help?

hahmadia replied:
Oops! My error. Thanks for pointing it out Scott. They do get promoted and demoted if the filter changes. Admins do not maintain their status if the filter changes. Is this correct behaviour? Or should active admins maintain their status if the filter changes/removed?

sbishel replied:
Yes, the admins should be promoted and demoted if the filter changes. The implementation is correct, just need to fix the text.

hahmadia replied:
Updated this text as provided by Dennis below!

sbishel commented on components/admin_console/admin_definition.jsx:
```diff
@@ -3099,6 +3114,18 @@ const AdminDefinition = {
+                        placeholder: t('admin.saml.adminAttrEx'),
+                        placeholder_default: 'E.g.: "usertype=Admin" or "isAdmin=true"',
+                        help_text: t('admin.saml.adminAttrDesc'),
+                        help_text_default: '(Optional) The attribute in the SAML Assertion that will be used to apply a admin role to users in Mattermost.\n \nNote: If this attribute is removed/changed from your admin user in SAML and the user is still active, they will not be demoted to a member and will retain their Admin role. Admins can be demoted in **System Console > User Management**.\n \n ...
```
@thefactremains Help with this one as well. The user will retain Admin role until they next login.

thefactremains replied:
> Admin Filter:
> (Optional) Enter an AD/LDAP filter to use for designating System Admins. The users selected by the query will have access to your Mattermost server as System Admins. By default, System Admins have complete access to the Mattermost System Console. 
> 
> Existing members that are identified by this attribute will be promoted from member to System Admin upon next login. The next login is based upon Session lengths set in System Console > Session Lengths. It is highly recommend to manually demote users to members in System Console > User Management to ensure access is restricted immediately.
> 
> Note: If this filter is removed/changed, System Admins that were promoted via this filter will be demoted to  members and will not retain access to the System Console. When this filter is not in use, System Admins can be manually promoted/demoted in System Console > User Management.

thefactremains replied:
@justinegeffen The above comment describes the functionality of this filter. Not sure where you want to document it. Let's discuss on our weekly?

hahmadia commented:
@srkgupta 
I have addressed the caching problem. I also added a `Enable Admin Filter` and `Enable Admin Attribute` for LDAP and SAML respectively.  Just to note how the workflow goes

```

sbishel commented:
Just one clarification on the note above. We don't distinguish between admins promoted by the filter and admins manually promoted. Even if a user is promoted to admin manually, they will be demoted on next login if they don't match the Admin filter/attribute and the option is enabled.

srkgupta commented on components/admin_console/admin_definition.jsx:
```diff
@@ -2454,6 +2454,34 @@ const AdminDefinition = {
+                        placeholder: t('admin.ldap.adminFilterEx'),
+                        placeholder_default: 'E.g.: "(objectClass=admins)"',
+                        isDisabled: it.either(
+                            it.configIsFalse('LdapSettings', 'EnableAdminFilter'),
```
Currently the Admin Filter attribute is not automatically enabled when we set "Enable Admin Filter" to true. We need to confirm the changes by clicking on Save button and only then this field gets enabled. Can we instead just check the state, something like this:
`it.StateIsFalse('LdapSettings.EnableAdminFilter')`

hahmadia commented:
/update-branch

mattermod commented:
Will try to auto merge this PR once all tests and checks are passing. This might take up to an hour.

mattermod commented:
Trying to auto merge this PR.

mattermod commented:
Pull Request successfully merged
SHA: 8eeeb52ca9f7663a7ba67da02fbeaa1b0d20276c

justinegeffen commented:
Documentation PRs: https://github.com/mattermost/docs/pull/3359 and https://github.com/mattermost/docs/pull/3355.
